rustls/src/webpki/verify.rs

```diff
@@ -457,6 +508,40 @@ impl<'a> TryFrom<&'a CertificateDer<'a>> for ParsedCertificate<'a> {
     }
 }
 
+/// Verify that the end-entity certificate `end_entity` is a valid server cert
+/// and chains to at least one of the trust anchors in the `roots` [RootCertStore].
+///
+/// `intermediates` contains all certificates other than `end_entity` that
+/// were sent as part of the server's `Certificate` message. It is in the
+/// same order that the server sent them and may be empty.
+///
+/// `revocation` controls how revocation checking is performed, if at all.
+///
+/// This function exists to be used by [`verify_server_cert_signed_by_trust_anchor`],
+/// and differs only in providing a `Option<webpki::RevocationOptions>` argument. We
+/// can't include this argument in `verify_server_cert_signed_by_trust_anchor` because
+/// it will leak the webpki types into Rustls' public API.
+pub(crate) fn verify_server_cert_signed_by_trust_anchor_impl(
+    cert: &ParsedCertificate,
+    roots: &RootCertStore,
+    intermediates: &[CertificateDer<'_>],
+    revocation: Option<webpki::RevocationOptions>,
+    now: UnixTime,
+    supported_algs: &[&dyn SignatureVerificationAlgorithm],
+) -> Result<(), Error> {
+    cert.0
+        .verify_for_usage(
+            supported_algs,
+            &roots.roots,
+            intermediates,
+            now,
+            webpki::KeyUsage::server_auth(),
+            revocation,
+        )
+        .map_err(pki_error)
```

Comment:
I tend to dislike chaining combinators that operate on different variants -- maybe try binding `result` and using an explicit `match` for that, see what you think?

Reply:
That's fair. I think the explicit `match` reads better :+1: Updated.